### PR TITLE
Fix incorrect use of BN_CTX API

### LINKED ISSUE #13812: Incorrect usage of BN_CONTEXT API
By modeling the issue fixed by commit 0e8b6c97ba7ac37f5e92f6a24d128b04b5336388 with codeql I was able to find some variants of the pattern .

Bug 1:

https://github.com/openssl/openssl/blob/9311d0c471ca2eaa259e8c1bbbeb7c46394c7ba2/crypto/bn/bn_prime.c#L149

If `mods = OPENSSL_zalloc(sizeof(*mods) * NUMPRIMES);` fails to allocate memory it will execute a goto that will perform `BN_CTX_end(ctx);` on the context without a matching `BN_CTX_start`.

Bug 2:

https://github.com/openssl/openssl/blob/605856d72cbd4720ad9ccb20c8fd5afbee2ad1a4/crypto/bn/bn_sqrt.c#L31

If `ret = BN_new();` fails to allocate memory it will execute a goto that will perform `BN_CTX_end(ctx);` on the context without a matching `BN_CTX_start`.

This can happen twice in the function so both cases would need to be addressed.

Bug 3:

https://github.com/openssl/openssl/blob/42619397eb5db1a77d077250b0841b9c9f2b8984/crypto/bn/bn_x931p.c#L177

If `BN_priv_rand_ex(Xp, nbits, BN_RAND_TOP_TWO, BN_RAND_BOTTOM_ANY, ctx))` fails it will execute a goto that will perform `BN_CTX_end(ctx);` on the context without a matching `BN_CTX_start`.

Bug 4:

https://github.com/openssl/openssl/blob/9311d0c471ca2eaa259e8c1bbbeb7c46394c7ba2/crypto/ec/ec_mult.c#L850

If `generator = EC_GROUP_get0_generator(group);` is null, it will execute a goto that will perform `BN_CTX_end(ctx);` on the context without a matching `BN_CTX_start`.

I've already fixed the issues and am creating the issue to refer to it in the PR.

### PR DESCRIPTION
**Pull request by agustingianni**
In some edge cases BN_CTX_end was being called without first calling
BN_CTX_start. This creates a situation where the state of the big
number allocator is corrupted and may lead to crashes.

Fixes #13812

### CODE REVIEW

**Review by mattcaswell — APPROVED** [review]
I think this may stray beyond our definition of trivial (which is generally reserved for very simple fixes, and typos etc). Can you submit a CLA?

https://www.openssl.org/policies/cla.html

**Comment by t8m** [comment]
I agree with Matt that this is beyond the trivial limit.

So please remove the CLA: trivial from the commit message and please also add the Fixes line to the commit message instead.

**Comment by agustingianni** [comment]
Hello @mattcaswell and @t8m as requested I have signed and sent the CLA to the openssl foundation. I have also removed the trivial tag and added the fixes line to the commit message.

Lemme know if there is anything else I need to do.

Thank you!

**Comment by mattcaswell** [comment]
Close/reopen to kick CLA bot

**Comment by mattcaswell** [comment]
CLA received - thank you! My approval stands.

Needs a second review.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master. Thank you for the contribution?

As this is something that should be fixed on the 1.1.1 branch but the cherry pick will not apply cleanly, would you like to submit a PR against the 1.1.1 branch, please?

**Pull request closed without merging**
